// @backstage/plugin-api-docs/dist/index.esm.js
import '@backstage/core';
import '@backstage/plugin-catalog-react';
import '@material-ui/lab';
import 'react';
export { a as ApiDefinitionCard, h as ApiExplorerPage, b as ApiTypeTitle, n as AsyncApiDefinitionWidget, C as ConsumedApisCard, c as ConsumingComponentsCard, E as EntityApiDefinitionCard, i as EntityConsumedApisCard, j as EntityConsumingComponentsCard, k as EntityHasApisCard, l as EntityProvidedApisCard, m as EntityProvidingComponentsCard, H as HasApisCard, O as OpenApiDefinitionWidget, o as PlainApiDefinitionWidget, P as ProvidedApisCard, e as ProvidingComponentsCard, f as apiDocsConfigRef, g as apiDocsPlugin, d as defaultDefinitionWidgets, g as plugin } from './esm/index-70fbe4d9.esm.js';
import '@asyncapi/react-component';
import '@asyncapi/react-component/lib/styles/fiori.css';
import '@material-ui/core/styles';
import 'graphiql/graphiql.css';
import 'graphql';
import 'swagger-ui-react';
import 'swagger-ui-react/swagger-ui.css';
import '@backstage/catalog-model';
import '@material-ui/core';
import 'react-router-dom';
import 'react-use';
