// @backstage/cli/dist/cjs/print-cef02c14.cjs.js
'use strict';

require('commander');
require('chalk');
require('./index-5c3253a4.cjs.js');
require('fs-extra');
require('@backstage/cli-common');
require('@backstage/config/package.json');
require('@backstage/config-loader');
var config$1 = require('@backstage/config');
var config = require('./config-f0226821.cjs.js');
var yaml = require('yaml');

var print = async (cmd) => {
  const {schema, appConfigs} = await config.loadCliConfig({
    args: cmd.config,
    fromPackage: cmd.package,
    mockEnv: cmd.lax
  });
  const visibility = getVisibilityOption(cmd);
  const data = serializeConfigData(appConfigs, schema, visibility);
  if (cmd.format === "json") {
    process.stdout.write(`${JSON.stringify(data, null, 2)}
`);
  } else {
    process.stdout.write(`${yaml.stringify(data)}
`);
  }
};
function getVisibilityOption(cmd) {
  if (cmd.frontend && cmd.withSecrets) {
    throw new Error("Not allowed to combine frontend and secret config");
  }
  if (cmd.frontend) {
    return "frontend";
  } else if (cmd.withSecrets) {
    return "secret";
  }
  return "backend";
}
function serializeConfigData(appConfigs, schema, visibility) {
  if (visibility === "frontend") {
    const frontendConfigs = schema.process(appConfigs, {
      visibility: ["frontend"]
    });
    return config$1.ConfigReader.fromConfigs(frontendConfigs).get();
  } else if (visibility === "secret") {
    return config$1.ConfigReader.fromConfigs(appConfigs).get();
  }
  const sanitizedConfigs = schema.process(appConfigs, {
    valueTransform: (value, context) => context.visibility === "secret" ? "<secret>" : value
  });
  return config$1.ConfigReader.fromConfigs(sanitizedConfigs).get();
}

exports.default = print;
