// @backstage/cli/dist/cjs/tasks-62a62154.cjs.js
'use strict';

var chalk = require('chalk');
var index = require('./index-5c3253a4.cjs.js');
var fs = require('fs-extra');
var path = require('path');
var handlebars = require('handlebars');
var ora = require('ora');
var recursive = require('recursive-readdir');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var chalk__default = /*#__PURE__*/_interopDefaultLegacy(chalk);
var fs__default = /*#__PURE__*/_interopDefaultLegacy(fs);
var path__default = /*#__PURE__*/_interopDefaultLegacy(path);
var handlebars__default = /*#__PURE__*/_interopDefaultLegacy(handlebars);
var ora__default = /*#__PURE__*/_interopDefaultLegacy(ora);
var recursive__default = /*#__PURE__*/_interopDefaultLegacy(recursive);

const TEAM_ID_RE = /^@[-\w]+\/[-\w]+$/;
const USER_ID_RE = /^@[-\w]+$/;
const EMAIL_RE = /^[^@]+@[-.\w]+\.[-\w]+$/i;
const DEFAULT_OWNER = "@backstage/maintainers";
async function getCodeownersFilePath(rootDir) {
  const paths = [
    path__default['default'].join(rootDir, ".github", "CODEOWNERS"),
    path__default['default'].join(rootDir, ".gitlab", "CODEOWNERS"),
    path__default['default'].join(rootDir, "docs", "CODEOWNERS"),
    path__default['default'].join(rootDir, "CODEOWNERS")
  ];
  for (const p of paths) {
    if (await fs__default['default'].pathExists(p)) {
      return p;
    }
  }
  return void 0;
}
function isValidSingleOwnerId(id) {
  if (!id || typeof id !== "string") {
    return false;
  }
  return TEAM_ID_RE.test(id) || USER_ID_RE.test(id) || EMAIL_RE.test(id);
}
function parseOwnerIds(spaceSeparatedOwnerIds) {
  if (!spaceSeparatedOwnerIds || typeof spaceSeparatedOwnerIds !== "string") {
    return void 0;
  }
  const ids = spaceSeparatedOwnerIds.split(" ").filter(Boolean);
  if (!ids.every(isValidSingleOwnerId)) {
    return void 0;
  }
  return ids;
}
async function addCodeownersEntry(codeownersFilePath, ownedPath, ownerIds) {
  const allLines = (await fs__default['default'].readFile(codeownersFilePath, "utf8")).split("\n");
  const commentLines = [];
  for (const line of allLines) {
    if (line[0] !== "#") {
      break;
    }
    commentLines.push(line);
  }
  const oldDeclarationEntries = allLines.filter((line) => line[0] !== "#").map((line) => line.split(/\s+/).filter(Boolean)).filter((tokens) => tokens.length >= 2).map((tokens) => ({
    ownedPath: tokens[0],
    ownerIds: tokens.slice(1)
  }));
  const newDeclarationEntries = oldDeclarationEntries.filter((entry) => entry.ownedPath !== "*").concat([{ownedPath, ownerIds}]).sort((l1, l2) => l1.ownedPath.localeCompare(l2.ownedPath));
  newDeclarationEntries.unshift({
    ownedPath: "*",
    ownerIds: [DEFAULT_OWNER]
  });
  const longestOwnedPath = newDeclarationEntries.reduce((length, entry) => Math.max(length, entry.ownedPath.length), 0);
  const newDeclarationLines = newDeclarationEntries.map((entry) => {
    const entryPath = entry.ownedPath + " ".repeat(longestOwnedPath - entry.ownedPath.length);
    return [entryPath, ...entry.ownerIds].join(" ");
  });
  const newLines = [...commentLines, "", ...newDeclarationLines, ""];
  await fs__default['default'].writeFile(codeownersFilePath, newLines.join("\n"), "utf8");
}

const TASK_NAME_MAX_LENGTH = 14;
class Task {
  static log(name = "") {
    process.stdout.write(`${chalk__default['default'].green(name)}
`);
  }
  static error(message = "") {
    process.stdout.write(`
${chalk__default['default'].red(message)}

`);
  }
  static section(name) {
    const title = chalk__default['default'].green(`${name}:`);
    process.stdout.write(`
 ${title}
`);
  }
  static exit(code = 0) {
    process.exit(code);
  }
  static async forItem(task, item, taskFunc) {
    const paddedTask = chalk__default['default'].green(task.padEnd(TASK_NAME_MAX_LENGTH));
    const spinner = ora__default['default']({
      prefixText: chalk__default['default'].green(`  ${paddedTask}${chalk__default['default'].cyan(item)}`),
      spinner: "arc",
      color: "green"
    }).start();
    try {
      await taskFunc();
      spinner.succeed();
    } catch (error) {
      spinner.fail();
      throw error;
    }
  }
}
async function templatingTask(templateDir, destinationDir, context, versions) {
  const files = await recursive__default['default'](templateDir).catch((error) => {
    throw new Error(`Failed to read template directory: ${error.message}`);
  });
  const isMonoRepo = await fs__default['default'].pathExists(index.paths.resolveTargetRoot("lerna.json"));
  for (const file of files) {
    const destinationFile = file.replace(templateDir, destinationDir);
    await fs__default['default'].ensureDir(path.dirname(destinationFile));
    if (file.endsWith(".hbs")) {
      await Task.forItem("templating", path.basename(file), async () => {
        const destination = destinationFile.replace(/\.hbs$/, "");
        const template = await fs__default['default'].readFile(file);
        const compiled = handlebars__default['default'].compile(template.toString());
        const contents = compiled({name: path.basename(destination), ...context}, {
          helpers: {
            version(name) {
              if (versions[name]) {
                return versions[name];
              }
              throw new Error(`No version available for package ${name}`);
            }
          }
        });
        await fs__default['default'].writeFile(destination, contents).catch((error) => {
          throw new Error(`Failed to create file: ${destination}: ${error.message}`);
        });
      });
    } else {
      if (isMonoRepo && file.match("tsconfig.json")) {
        continue;
      }
      await Task.forItem("copying", path.basename(file), async () => {
        await fs__default['default'].copyFile(file, destinationFile).catch((error) => {
          const destination = destinationFile;
          throw new Error(`Failed to copy file to ${destination} : ${error.message}`);
        });
      });
    }
  }
}

exports.Task = Task;
exports.addCodeownersEntry = addCodeownersEntry;
exports.getCodeownersFilePath = getCodeownersFilePath;
exports.parseOwnerIds = parseOwnerIds;
exports.templatingTask = templatingTask;
