// @backstage/plugin-catalog/dist/esm/index-10f59334.esm.js
import { EDIT_URL_ANNOTATION, VIEW_URL_ANNOTATION, SOURCE_LOCATION_ANNOTATION, parseLocationReference, RELATION_PART_OF, RELATION_OWNED_BY, RELATION_PROVIDES_API, RELATION_CONSUMES_API, ENTITY_DEFAULT_NAMESPACE, ORIGIN_LOCATION_ANNOTATION, getEntityName } from '@backstage/catalog-model';
import { useApi, HeaderIconLinkRow, alertApiRef, configApiRef, Progress, ResponseErrorPanel, attachComponentData, Page, Header, RoutedTabs, Content, HeaderLabel, HeaderTabs, WarningPanel, Link, OverflowTooltip, CodeSnippet, Table, createExternalRouteRef, useRouteRef, ContentHeader, SupportButton, createPlugin, createApiFactory, discoveryApiRef, identityApiRef, createRoutableExtension, createComponentExtension } from '@backstage/core';
import { scmIntegrationsApiRef } from '@backstage/integration-react';
import { getEntityRelations, EntityRefLinks, useEntity, useStarredEntities, catalogApiRef, EntityRefLink, useEntityCompoundName, EntityContext, useEntityListProvider, formatEntityRefTitle, EntityListProvider, EntityKindPicker, EntityTypePicker, UserListPicker, EntityTagPicker, useEntityFromUrl, rootRoute, entityRoute, catalogRouteRef, entityRouteRef } from '@backstage/plugin-catalog-react';
import { makeStyles, Typography, Grid, Chip, Card, CardHeader, IconButton, Divider, CardContent, MenuItem, ListItemIcon, ListItemText, Popover, MenuList, withStyles, Tooltip, Dialog, DialogTitle, DialogContent, DialogActions, Button, Box, DialogContentText, Link as Link$2 } from '@material-ui/core';
import DocsIcon from '@material-ui/icons/Description';
import EditIcon from '@material-ui/icons/Edit';
import ExtensionIcon from '@material-ui/icons/Extension';
import React, { useState, useCallback, useContext, Children, isValidElement, Fragment, useMemo } from 'react';
import CodeIcon from '@material-ui/icons/Code';
import GitHubIcon from '@material-ui/icons/GitHub';
import { Alert as Alert$1 } from '@material-ui/lab';
import { useNavigate, useParams, Navigate, matchRoutes, useRoutes, Routes, Route as Route$1 } from 'react-router';
import { makeStyles as makeStyles$1 } from '@material-ui/core/styles';
import Cancel from '@material-ui/icons/Cancel';
import MoreVert from '@material-ui/icons/MoreVert';
import StarBorder from '@material-ui/icons/StarBorder';
import Star from '@material-ui/icons/Star';
import Alert from '@material-ui/lab/Alert';
import { useAsync } from 'react-use';
import { Helmet } from 'react-helmet';
import OpenInNew from '@material-ui/icons/OpenInNew';
import { capitalize } from 'lodash';
import { Link as Link$1 } from 'react-router-dom';
import IlloSvgUrl from '../components/EntityNotFound/Illo/illo.svg';
import { CatalogClient } from '@backstage/catalog-client';

function getEntityMetadataViewUrl(entity) {
  var _a;
  return (_a = entity.metadata.annotations) == null ? void 0 : _a[VIEW_URL_ANNOTATION];
}
function getEntityMetadataEditUrl(entity) {
  var _a;
  return (_a = entity.metadata.annotations) == null ? void 0 : _a[EDIT_URL_ANNOTATION];
}

function getEntitySourceLocation(entity, scmIntegrationsApi) {
  var _a;
  const sourceLocation = (_a = entity.metadata.annotations) == null ? void 0 : _a[SOURCE_LOCATION_ANNOTATION];
  if (!sourceLocation) {
    return void 0;
  }
  try {
    const sourceLocationRef = parseLocationReference(sourceLocation);
    const integration = scmIntegrationsApi.byUrl(sourceLocationRef.target);
    return {
      locationTargetUrl: sourceLocationRef.target,
      integrationType: integration == null ? void 0 : integration.type
    };
  } catch {
    return void 0;
  }
}

const useStyles = makeStyles((theme) => ({
  value: {
    fontWeight: "bold",
    overflow: "hidden",
    lineHeight: "24px",
    wordBreak: "break-word"
  },
  label: {
    color: theme.palette.text.secondary,
    textTransform: "uppercase",
    fontSize: "10px",
    fontWeight: "bold",
    letterSpacing: 0.5,
    overflow: "hidden",
    whiteSpace: "nowrap"
  }
}));
const AboutField = ({label, value, gridSizes, children}) => {
  const classes = useStyles();
  const content = React.Children.count(children) ? children : /* @__PURE__ */ React.createElement(Typography, {
    variant: "body2",
    className: classes.value
  }, value || `unknown`);
  return /* @__PURE__ */ React.createElement(Grid, {
    item: true,
    ...gridSizes
  }, /* @__PURE__ */ React.createElement(Typography, {
    variant: "subtitle2",
    className: classes.label
  }, label), content);
};

const useStyles$1 = makeStyles({
  description: {
    wordBreak: "break-word"
  }
});
const AboutContent = ({entity}) => {
  var _a, _b, _c, _d;
  const classes = useStyles$1();
  const isSystem = entity.kind.toLocaleLowerCase("en-US") === "system";
  const isDomain = entity.kind.toLocaleLowerCase("en-US") === "domain";
  const isResource = entity.kind.toLocaleLowerCase("en-US") === "resource";
  const isComponent = entity.kind.toLocaleLowerCase("en-US") === "component";
  const partOfSystemRelations = getEntityRelations(entity, RELATION_PART_OF, {
    kind: "system"
  });
  const partOfComponentRelations = getEntityRelations(entity, RELATION_PART_OF, {
    kind: "component"
  });
  const partOfDomainRelations = getEntityRelations(entity, RELATION_PART_OF, {
    kind: "domain"
  });
  const ownedByRelations = getEntityRelations(entity, RELATION_OWNED_BY);
  return /* @__PURE__ */ React.createElement(Grid, {
    container: true
  }, /* @__PURE__ */ React.createElement(AboutField, {
    label: "Description",
    gridSizes: {xs: 12}
  }, /* @__PURE__ */ React.createElement(Typography, {
    variant: "body2",
    paragraph: true,
    className: classes.description
  }, ((_a = entity == null ? void 0 : entity.metadata) == null ? void 0 : _a.description) || "No description")), /* @__PURE__ */ React.createElement(AboutField, {
    label: "Owner",
    gridSizes: {xs: 12, sm: 6, lg: 4}
  }, /* @__PURE__ */ React.createElement(EntityRefLinks, {
    entityRefs: ownedByRelations,
    defaultKind: "group"
  })), isSystem && /* @__PURE__ */ React.createElement(AboutField, {
    label: "Domain",
    value: "No Domain",
    gridSizes: {xs: 12, sm: 6, lg: 4}
  }, /* @__PURE__ */ React.createElement(EntityRefLinks, {
    entityRefs: partOfDomainRelations,
    defaultKind: "domain"
  })), !isSystem && !isDomain && /* @__PURE__ */ React.createElement(AboutField, {
    label: "System",
    value: "No System",
    gridSizes: {xs: 12, sm: 6, lg: 4}
  }, /* @__PURE__ */ React.createElement(EntityRefLinks, {
    entityRefs: partOfSystemRelations,
    defaultKind: "system"
  })), isComponent && partOfComponentRelations.length > 0 && /* @__PURE__ */ React.createElement(AboutField, {
    label: "Parent Component",
    value: "No Parent Component",
    gridSizes: {xs: 12, sm: 6, lg: 4}
  }, /* @__PURE__ */ React.createElement(EntityRefLinks, {
    entityRefs: partOfComponentRelations,
    defaultKind: "component"
  })), !isSystem && !isDomain && /* @__PURE__ */ React.createElement(AboutField, {
    label: "Type",
    value: (_b = entity == null ? void 0 : entity.spec) == null ? void 0 : _b.type,
    gridSizes: {xs: 12, sm: 6, lg: 4}
  }), !isSystem && !isDomain && !isResource && /* @__PURE__ */ React.createElement(AboutField, {
    label: "Lifecycle",
    value: (_c = entity == null ? void 0 : entity.spec) == null ? void 0 : _c.lifecycle,
    gridSizes: {xs: 12, sm: 6, lg: 4}
  }), /* @__PURE__ */ React.createElement(AboutField, {
    label: "Tags",
    value: "No Tags",
    gridSizes: {xs: 12, sm: 6, lg: 4}
  }, (((_d = entity == null ? void 0 : entity.metadata) == null ? void 0 : _d.tags) || []).map((t) => /* @__PURE__ */ React.createElement(Chip, {
    key: t,
    size: "small",
    label: t
  }))));
};

const ScmIntegrationIcon = ({type}) => {
  switch (type) {
    case "github":
      return /* @__PURE__ */ React.createElement(GitHubIcon, null);
    default:
      return /* @__PURE__ */ React.createElement(CodeIcon, null);
  }
};

const useStyles$2 = makeStyles({
  gridItemCard: {
    display: "flex",
    flexDirection: "column",
    height: "calc(100% - 10px)",
    marginBottom: "10px"
  },
  fullHeightCard: {
    display: "flex",
    flexDirection: "column",
    height: "100%"
  },
  gridItemCardContent: {
    flex: 1
  },
  fullHeightCardContent: {
    flex: 1
  }
});
function AboutCard({variant}) {
  var _a;
  const classes = useStyles$2();
  const {entity} = useEntity();
  const scmIntegrationsApi = useApi(scmIntegrationsApiRef);
  const entitySourceLocation = getEntitySourceLocation(entity, scmIntegrationsApi);
  const entityMetadataEditUrl = getEntityMetadataEditUrl(entity);
  const providesApiRelations = getEntityRelations(entity, RELATION_PROVIDES_API);
  const consumesApiRelations = getEntityRelations(entity, RELATION_CONSUMES_API);
  const hasApis = providesApiRelations.length > 0 || consumesApiRelations.length > 0;
  const viewInSource = {
    label: "View Source",
    disabled: !entitySourceLocation,
    icon: /* @__PURE__ */ React.createElement(ScmIntegrationIcon, {
      type: entitySourceLocation == null ? void 0 : entitySourceLocation.integrationType
    }),
    href: entitySourceLocation == null ? void 0 : entitySourceLocation.locationTargetUrl
  };
  const viewInTechDocs = {
    label: "View TechDocs",
    disabled: !((_a = entity.metadata.annotations) == null ? void 0 : _a["backstage.io/techdocs-ref"]),
    icon: /* @__PURE__ */ React.createElement(DocsIcon, null),
    href: `/docs/${entity.metadata.namespace || ENTITY_DEFAULT_NAMESPACE}/${entity.kind}/${entity.metadata.name}`
  };
  const viewApi = {
    title: hasApis ? "" : "No APIs available",
    label: "View API",
    disabled: !hasApis,
    icon: /* @__PURE__ */ React.createElement(ExtensionIcon, null),
    href: "api"
  };
  let cardClass = "";
  if (variant === "gridItem") {
    cardClass = classes.gridItemCard;
  } else if (variant === "fullHeight") {
    cardClass = classes.fullHeightCard;
  }
  let cardContentClass = "";
  if (variant === "gridItem") {
    cardContentClass = classes.gridItemCardContent;
  } else if (variant === "fullHeight") {
    cardContentClass = classes.fullHeightCardContent;
  }
  return /* @__PURE__ */ React.createElement(Card, {
    className: cardClass
  }, /* @__PURE__ */ React.createElement(CardHeader, {
    title: "About",
    action: /* @__PURE__ */ React.createElement(IconButton, {
      "aria-label": "Edit",
      disabled: !entityMetadataEditUrl,
      title: "Edit Metadata",
      onClick: () => {
        window.open(entityMetadataEditUrl != null ? entityMetadataEditUrl : "#", "_blank");
      }
    }, /* @__PURE__ */ React.createElement(EditIcon, null)),
    subheader: /* @__PURE__ */ React.createElement(HeaderIconLinkRow, {
      links: [viewInSource, viewInTechDocs, viewApi]
    })
  }), /* @__PURE__ */ React.createElement(Divider, null), /* @__PURE__ */ React.createElement(CardContent, {
    className: cardContentClass
  }, /* @__PURE__ */ React.createElement(AboutContent, {
    entity
  })));
}

const useStyles$3 = makeStyles$1({
  button: {
    color: "white"
  }
});
const EntityContextMenu = ({
  UNSTABLE_extraContextMenuItems,
  onUnregisterEntity
}) => {
  const [anchorEl, setAnchorEl] = useState();
  const classes = useStyles$3();
  const onOpen = (event) => {
    setAnchorEl(event.currentTarget);
  };
  const onClose = () => {
    setAnchorEl(void 0);
  };
  const extraItems = UNSTABLE_extraContextMenuItems && [
    ...UNSTABLE_extraContextMenuItems.map((item) => /* @__PURE__ */ React.createElement(MenuItem, {
      key: item.title,
      onClick: () => {
        onClose();
        item.onClick();
      }
    }, /* @__PURE__ */ React.createElement(ListItemIcon, null, /* @__PURE__ */ React.createElement(item.Icon, {
      fontSize: "small"
    })), /* @__PURE__ */ React.createElement(ListItemText, {
      primary: item.title
    }))),
    /* @__PURE__ */ React.createElement(Divider, {
      key: "the divider is here!"
    })
  ];
  return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(IconButton, {
    "aria-label": "more",
    "aria-controls": "long-menu",
    "aria-haspopup": "true",
    onClick: onOpen,
    "data-testid": "menu-button",
    className: classes.button
  }, /* @__PURE__ */ React.createElement(MoreVert, null)), /* @__PURE__ */ React.createElement(Popover, {
    open: Boolean(anchorEl),
    onClose,
    anchorEl,
    anchorOrigin: {vertical: "bottom", horizontal: "right"},
    transformOrigin: {vertical: "top", horizontal: "right"}
  }, /* @__PURE__ */ React.createElement(MenuList, null, extraItems, /* @__PURE__ */ React.createElement(MenuItem, {
    onClick: () => {
      onClose();
      onUnregisterEntity();
    }
  }, /* @__PURE__ */ React.createElement(ListItemIcon, null, /* @__PURE__ */ React.createElement(Cancel, {
    fontSize: "small"
  })), /* @__PURE__ */ React.createElement(ListItemText, {
    primary: "Unregister entity"
  })))));
};

const YellowStar = withStyles({
  root: {
    color: "#f3ba37"
  }
})(Star);
const favouriteEntityTooltip = (isStarred) => isStarred ? "Remove from favorites" : "Add to favorites";
const favouriteEntityIcon = (isStarred) => isStarred ? /* @__PURE__ */ React.createElement(YellowStar, null) : /* @__PURE__ */ React.createElement(StarBorder, null);
const FavouriteEntity = (props) => {
  const {toggleStarredEntity, isStarredEntity} = useStarredEntities();
  const isStarred = isStarredEntity(props.entity);
  return /* @__PURE__ */ React.createElement(IconButton, {
    color: "inherit",
    ...props,
    onClick: () => toggleStarredEntity(props.entity)
  }, /* @__PURE__ */ React.createElement(Tooltip, {
    title: favouriteEntityTooltip(isStarred)
  }, favouriteEntityIcon(isStarred)));
};

function useUnregisterEntityDialogState(entity) {
  var _a;
  const catalogApi = useApi(catalogApiRef);
  const locationRef = (_a = entity.metadata.annotations) == null ? void 0 : _a[ORIGIN_LOCATION_ANNOTATION];
  const uid = entity.metadata.uid;
  const isBootstrap = locationRef === "bootstrap:bootstrap";
  const prerequisites = useAsync(async () => {
    const locationPromise = catalogApi.getOriginLocationByEntity(entity);
    let colocatedEntitiesPromise;
    if (!locationRef) {
      colocatedEntitiesPromise = Promise.resolve([]);
    } else {
      const locationAnnotationFilter = `metadata.annotations.${ORIGIN_LOCATION_ANNOTATION}`;
      colocatedEntitiesPromise = catalogApi.getEntities({
        filter: {[locationAnnotationFilter]: locationRef},
        fields: [
          "kind",
          "metadata.uid",
          "metadata.name",
          "metadata.namespace"
        ]
      }).then((response) => response.items);
    }
    return Promise.all([locationPromise, colocatedEntitiesPromise]).then(([location2, colocatedEntities2]) => ({
      location: location2,
      colocatedEntities: colocatedEntities2
    }));
  }, [catalogApi, entity]);
  const unregisterLocation = useCallback(async function unregisterLocationFn() {
    const {location: location2, colocatedEntities: colocatedEntities2} = prerequisites.value;
    await catalogApi.removeLocationById(location2.id);
    await Promise.allSettled(colocatedEntities2.map((e) => catalogApi.removeEntityByUid(e.metadata.uid)));
  }, [catalogApi, prerequisites]);
  const deleteEntity = useCallback(async function deleteEntityFn() {
    await catalogApi.removeEntityByUid(uid);
  }, [catalogApi, uid]);
  if (isBootstrap) {
    return {type: "bootstrap", location: locationRef, deleteEntity};
  }
  const {loading, error, value} = prerequisites;
  if (loading) {
    return {type: "loading"};
  } else if (error) {
    return {type: "error", error};
  }
  const {location, colocatedEntities} = value;
  if (!location) {
    return {type: "only-delete", deleteEntity};
  }
  return {
    type: "unregister",
    location: locationRef,
    colocatedEntities: colocatedEntities.map(getEntityName),
    unregisterLocation,
    deleteEntity
  };
}

const useStyles$4 = makeStyles({
  advancedButton: {
    fontSize: "0.7em"
  }
});
const Contents = ({
  entity,
  onConfirm
}) => {
  var _a;
  const alertApi = useApi(alertApiRef);
  const configApi = useApi(configApiRef);
  const classes = useStyles$4();
  const state = useUnregisterEntityDialogState(entity);
  const [showDelete, setShowDelete] = useState(false);
  const [busy, setBusy] = useState(false);
  const appTitle = (_a = configApi.getOptionalString("app.title")) != null ? _a : "Backstage";
  const onUnregister = useCallback(async function onUnregisterFn() {
    if ("unregisterLocation" in state) {
      setBusy(true);
      try {
        await state.unregisterLocation();
        onConfirm();
      } catch (err) {
        alertApi.post({message: err.message});
      } finally {
        setBusy(false);
      }
    }
  }, [alertApi, onConfirm, state]);
  const onDelete = useCallback(async function onDeleteFn() {
    if ("deleteEntity" in state) {
      setBusy(true);
      try {
        await state.deleteEntity();
        onConfirm();
      } catch (err) {
        alertApi.post({message: err.message});
      } finally {
        setBusy(false);
      }
    }
  }, [alertApi, onConfirm, state]);
  if (state.type === "loading") {
    return /* @__PURE__ */ React.createElement(Progress, null);
  }
  if (state.type === "error") {
    return /* @__PURE__ */ React.createElement(ResponseErrorPanel, {
      error: state.error
    });
  }
  if (state.type === "bootstrap") {
    return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(Alert, {
      severity: "info"
    }, 'You cannot unregister this entity, since it originates from a protected Backstage configuration (location "', state.location, '"). If you believe this is in error, please contact the ', appTitle, " ", "integrator."), /* @__PURE__ */ React.createElement(Box, {
      marginTop: 2
    }, !showDelete && /* @__PURE__ */ React.createElement(Button, {
      variant: "text",
      size: "small",
      color: "primary",
      className: classes.advancedButton,
      onClick: () => setShowDelete(true)
    }, "Advanced Options"), showDelete && /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(DialogContentText, null, "You have the option to delete the entity itself from the catalog. Note that this should only be done if you know that the catalog file has been deleted at, or moved from, its origin location. If that is not the case, the entity will reappear shortly as the next refresh round is performed by the catalog."), /* @__PURE__ */ React.createElement(Button, {
      variant: "contained",
      color: "secondary",
      disabled: busy,
      onClick: onDelete
    }, "Delete Entity"))));
  }
  if (state.type === "only-delete") {
    return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(DialogContentText, null, "This entity does not seem to originate from a location. You therefore only have the option to delete it outright from the catalog."), /* @__PURE__ */ React.createElement(Button, {
      variant: "contained",
      color: "secondary",
      disabled: busy,
      onClick: onDelete
    }, "Delete Entity"));
  }
  if (state.type === "unregister") {
    return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(DialogContentText, null, "This action will unregister the following entities:"), /* @__PURE__ */ React.createElement(DialogContentText, {
      component: "ul"
    }, state.colocatedEntities.map((e) => /* @__PURE__ */ React.createElement("li", {
      key: `${e.kind}:${e.namespace}/${e.name}`
    }, /* @__PURE__ */ React.createElement(EntityRefLink, {
      entityRef: e
    })))), /* @__PURE__ */ React.createElement(DialogContentText, null, "Located at the following location:"), /* @__PURE__ */ React.createElement(DialogContentText, {
      component: "ul"
    }, /* @__PURE__ */ React.createElement("li", null, state.location)), /* @__PURE__ */ React.createElement(DialogContentText, null, "To undo, just re-register the entity in ", appTitle, "."), /* @__PURE__ */ React.createElement(Box, {
      marginTop: 2
    }, /* @__PURE__ */ React.createElement(Button, {
      variant: "contained",
      color: "secondary",
      disabled: busy,
      onClick: onUnregister
    }, "Unregister Location"), !showDelete && /* @__PURE__ */ React.createElement(Box, {
      component: "span",
      marginLeft: 2
    }, /* @__PURE__ */ React.createElement(Button, {
      variant: "text",
      size: "small",
      color: "primary",
      className: classes.advancedButton,
      onClick: () => setShowDelete(true)
    }, "Advanced Options"))), showDelete && /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(Box, {
      paddingTop: 4,
      paddingBottom: 4
    }, /* @__PURE__ */ React.createElement(Divider, null)), /* @__PURE__ */ React.createElement(DialogContentText, null, "You also have the option to delete the entity itself from the catalog. Note that this should only be done if you know that the catalog file has been deleted at, or moved from, its origin location. If that is not the case, the entity will reappear shortly as the next refresh round is performed by the catalog."), /* @__PURE__ */ React.createElement(Button, {
      variant: "contained",
      color: "secondary",
      disabled: busy,
      onClick: onDelete
    }, "Delete Entity")));
  }
  return /* @__PURE__ */ React.createElement(Alert, {
    severity: "error"
  }, "Internal error: Unknown state");
};
const UnregisterEntityDialog = ({
  open,
  onConfirm,
  onClose,
  entity
}) => /* @__PURE__ */ React.createElement(Dialog, {
  open,
  onClose
}, /* @__PURE__ */ React.createElement(DialogTitle, {
  id: "responsive-dialog-title"
}, "Are you sure you want to unregister this entity?"), /* @__PURE__ */ React.createElement(DialogContent, null, /* @__PURE__ */ React.createElement(Contents, {
  entity,
  onConfirm
})), /* @__PURE__ */ React.createElement(DialogActions, null, /* @__PURE__ */ React.createElement(Button, {
  onClick: onClose,
  color: "primary"
}, "Cancel")));

const Route = () => null;
attachComponentData(Route, "core.gatherMountPoints", true);
function createSubRoutesFromChildren(childrenProps, entity) {
  const routeType = (/* @__PURE__ */ React.createElement(Route, {
    path: "",
    title: ""
  }, /* @__PURE__ */ React.createElement("div", null))).type;
  return Children.toArray(childrenProps).flatMap((child) => {
    if (!isValidElement(child)) {
      return [];
    }
    if (child.type === Fragment) {
      return createSubRoutesFromChildren(child.props.children, entity);
    }
    if (child.type !== routeType) {
      throw new Error("Child of EntityLayout must be an EntityLayout.Route");
    }
    const {path, title, children, if: condition, tabProps} = child.props;
    if (condition && entity && !condition(entity)) {
      return [];
    }
    return [{path, title, children, tabProps}];
  });
}
const EntityLayoutTitle = ({
  entity,
  title
}) => /* @__PURE__ */ React.createElement(Box, {
  display: "inline-flex",
  alignItems: "center",
  height: "1em"
}, title, entity && /* @__PURE__ */ React.createElement(FavouriteEntity, {
  entity
}));
const headerProps = (paramKind, paramNamespace, paramName, entity) => {
  var _a, _b, _c;
  const kind = (_a = paramKind != null ? paramKind : entity == null ? void 0 : entity.kind) != null ? _a : "";
  const namespace = (_b = paramNamespace != null ? paramNamespace : entity == null ? void 0 : entity.metadata.namespace) != null ? _b : "";
  const name = (_c = paramName != null ? paramName : entity == null ? void 0 : entity.metadata.name) != null ? _c : "";
  return {
    headerTitle: `${name}${namespace && namespace !== ENTITY_DEFAULT_NAMESPACE ? ` in ${namespace}` : ""}`,
    headerType: (() => {
      let t = kind.toLocaleLowerCase("en-US");
      if (entity && entity.spec && "type" in entity.spec) {
        t += " \u2014 ";
        t += entity.spec.type.toLocaleLowerCase("en-US");
      }
      return t;
    })()
  };
};
const EntityLabels = ({entity}) => {
  var _a;
  const ownedByRelations = getEntityRelations(entity, RELATION_OWNED_BY);
  return /* @__PURE__ */ React.createElement(React.Fragment, null, ownedByRelations.length > 0 && /* @__PURE__ */ React.createElement(HeaderLabel, {
    label: "Owner",
    value: /* @__PURE__ */ React.createElement(EntityRefLinks, {
      entityRefs: ownedByRelations,
      defaultKind: "Group",
      color: "inherit"
    })
  }), ((_a = entity.spec) == null ? void 0 : _a.lifecycle) && /* @__PURE__ */ React.createElement(HeaderLabel, {
    label: "Lifecycle",
    value: entity.spec.lifecycle
  }));
};
const EntityLayout = ({
  UNSTABLE_extraContextMenuItems,
  children
}) => {
  var _a, _b, _c;
  const {kind, namespace, name} = useEntityCompoundName();
  const {entity, loading, error} = useContext(EntityContext);
  const routes = createSubRoutesFromChildren(children, entity);
  const {headerTitle, headerType} = headerProps(kind, namespace, name, entity);
  const [confirmationDialogOpen, setConfirmationDialogOpen] = useState(false);
  const navigate = useNavigate();
  const cleanUpAfterRemoval = async () => {
    setConfirmationDialogOpen(false);
    navigate("/");
  };
  const showRemovalDialog = () => setConfirmationDialogOpen(true);
  return /* @__PURE__ */ React.createElement(Page, {
    themeId: (_c = (_b = (_a = entity == null ? void 0 : entity.spec) == null ? void 0 : _a.type) == null ? void 0 : _b.toString()) != null ? _c : "home"
  }, /* @__PURE__ */ React.createElement(Header, {
    title: /* @__PURE__ */ React.createElement(EntityLayoutTitle, {
      title: headerTitle,
      entity
    }),
    pageTitleOverride: headerTitle,
    type: headerType
  }, entity && /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(EntityLabels, {
    entity
  }), /* @__PURE__ */ React.createElement(EntityContextMenu, {
    UNSTABLE_extraContextMenuItems,
    onUnregisterEntity: showRemovalDialog
  }))), loading && /* @__PURE__ */ React.createElement(Progress, null), entity && /* @__PURE__ */ React.createElement(RoutedTabs, {
    routes
  }), error && /* @__PURE__ */ React.createElement(Content, null, /* @__PURE__ */ React.createElement(Alert$1, {
    severity: "error"
  }, error.toString())), /* @__PURE__ */ React.createElement(UnregisterEntityDialog, {
    open: confirmationDialogOpen,
    entity,
    onConfirm: cleanUpAfterRemoval,
    onClose: () => setConfirmationDialogOpen(false)
  }));
};
EntityLayout.Route = Route;

const getSelectedIndexOrDefault = (matchedRoute, tabs, defaultIndex = 0) => {
  if (!matchedRoute)
    return defaultIndex;
  const tabIndex = tabs.findIndex((t) => t.id === matchedRoute.route.path);
  return ~tabIndex ? tabIndex : defaultIndex;
};
const Tabbed = {
  Layout: ({children}) => {
    var _a, _b, _c;
    const routes = [];
    const tabs = [];
    const params = useParams();
    const navigate = useNavigate();
    React.Children.forEach(children, (child) => {
      if (!React.isValidElement(child)) {
        return;
      }
      if (child.type !== Tabbed.Content) {
        throw new Error("This component only accepts Content elements as direct children. Check the code of the EntityPage.");
      }
      const pathAndId = child.props.path;
      tabs.push({
        id: pathAndId,
        label: child.props.title
      });
      routes.push({
        path: pathAndId,
        element: child.props.element
      });
    });
    if (((_b = (_a = routes == null ? void 0 : routes[0]) == null ? void 0 : _a.path) != null ? _b : "") !== "")
      routes.push({
        path: "/*",
        element: /* @__PURE__ */ React.createElement(Navigate, {
          to: routes[0].path
        })
      });
    const [matchedRoute] = (_c = matchRoutes(routes, `/${params["*"]}`)) != null ? _c : [];
    const selectedIndex = getSelectedIndexOrDefault(matchedRoute, tabs);
    const currentTab = tabs[selectedIndex];
    const title = currentTab == null ? void 0 : currentTab.label;
    const onTabChange = (index) => navigate(tabs[index].id.replace(/\/\*$/, "").replace(/^\//, ""));
    const currentRouteElement = useRoutes(routes);
    if (!currentTab)
      return null;
    return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(HeaderTabs, {
      tabs,
      selectedIndex,
      onChange: onTabChange
    }), /* @__PURE__ */ React.createElement(Content, null, /* @__PURE__ */ React.createElement(Helmet, {
      title
    }), currentRouteElement));
  },
  Content: (_props) => null
};

const EntityPageTitle = ({
  entity,
  title
}) => /* @__PURE__ */ React.createElement(Box, {
  display: "inline-flex",
  alignItems: "center",
  height: "1em"
}, title, entity && /* @__PURE__ */ React.createElement(FavouriteEntity, {
  entity
}));
const EntityLabels$1 = ({entity}) => {
  var _a;
  const ownedByRelations = getEntityRelations(entity, RELATION_OWNED_BY);
  return /* @__PURE__ */ React.createElement(React.Fragment, null, ownedByRelations.length > 0 && /* @__PURE__ */ React.createElement(HeaderLabel, {
    label: "Owner",
    value: /* @__PURE__ */ React.createElement(EntityRefLinks, {
      entityRefs: ownedByRelations,
      defaultKind: "Group",
      color: "inherit"
    })
  }), ((_a = entity.spec) == null ? void 0 : _a.lifecycle) && /* @__PURE__ */ React.createElement(HeaderLabel, {
    label: "Lifecycle",
    value: entity.spec.lifecycle
  }));
};
const headerProps$1 = (kind, namespace, name, entity) => {
  return {
    headerTitle: `${name}${namespace && namespace !== ENTITY_DEFAULT_NAMESPACE ? ` in ${namespace}` : ""}`,
    headerType: (() => {
      let t = kind.toLocaleLowerCase("en-US");
      if (entity && entity.spec && "type" in entity.spec) {
        t += " \u2014 ";
        t += entity.spec.type.toLocaleLowerCase("en-US");
      }
      return t;
    })()
  };
};
const EntityPageLayout = ({
  children,
  UNSTABLE_extraContextMenuItems
}) => {
  var _a, _b, _c;
  const {kind, namespace, name} = useEntityCompoundName();
  const {entity, loading, error} = useContext(EntityContext);
  const {headerTitle, headerType} = headerProps$1(kind, namespace, name, entity);
  const [confirmationDialogOpen, setConfirmationDialogOpen] = useState(false);
  const navigate = useNavigate();
  const cleanUpAfterRemoval = async () => {
    setConfirmationDialogOpen(false);
    navigate("/");
  };
  const showRemovalDialog = () => setConfirmationDialogOpen(true);
  return /* @__PURE__ */ React.createElement(Page, {
    themeId: (_c = (_b = (_a = entity == null ? void 0 : entity.spec) == null ? void 0 : _a.type) == null ? void 0 : _b.toString()) != null ? _c : "home"
  }, /* @__PURE__ */ React.createElement(Header, {
    title: /* @__PURE__ */ React.createElement(EntityPageTitle, {
      title: headerTitle,
      entity
    }),
    pageTitleOverride: headerTitle,
    type: headerType
  }, entity && /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(EntityLabels$1, {
    entity
  }), /* @__PURE__ */ React.createElement(EntityContextMenu, {
    UNSTABLE_extraContextMenuItems,
    onUnregisterEntity: showRemovalDialog
  }))), loading && /* @__PURE__ */ React.createElement(Content, null, /* @__PURE__ */ React.createElement(Progress, null)), entity && /* @__PURE__ */ React.createElement(Tabbed.Layout, null, children), error && /* @__PURE__ */ React.createElement(Content, null, /* @__PURE__ */ React.createElement(ResponseErrorPanel, {
    error
  })), !loading && !error && !entity && /* @__PURE__ */ React.createElement(Content, null, /* @__PURE__ */ React.createElement(WarningPanel, {
    title: "Entity not found"
  }, "There is no ", kind, " with the requested", " ", /* @__PURE__ */ React.createElement(Link, {
    to: "https://backstage.io/docs/features/software-catalog/references"
  }, "kind, namespace, and name"), ".")), /* @__PURE__ */ React.createElement(UnregisterEntityDialog, {
    open: confirmationDialogOpen,
    entity,
    onConfirm: cleanUpAfterRemoval,
    onClose: () => setConfirmationDialogOpen(false)
  }));
};
EntityPageLayout.Content = Tabbed.Content;

function createNameColumn() {
  return {
    title: "Name",
    field: "resolved.name",
    highlight: true,
    render: ({entity}) => /* @__PURE__ */ React.createElement(EntityRefLink, {
      entityRef: entity,
      defaultKind: "Component"
    })
  };
}
function createSystemColumn() {
  return {
    title: "System",
    field: "resolved.partOfSystemRelationTitle",
    render: ({resolved}) => /* @__PURE__ */ React.createElement(EntityRefLinks, {
      entityRefs: resolved.partOfSystemRelations,
      defaultKind: "system"
    })
  };
}
function createOwnerColumn() {
  return {
    title: "Owner",
    field: "resolved.ownedByRelationsTitle",
    render: ({resolved}) => /* @__PURE__ */ React.createElement(EntityRefLinks, {
      entityRefs: resolved.ownedByRelations,
      defaultKind: "group"
    })
  };
}
function createSpecTypeColumn() {
  return {
    title: "Type",
    field: "entity.spec.type",
    hidden: true
  };
}
function createSpecLifecycleColumn() {
  return {
    title: "Lifecycle",
    field: "entity.spec.lifecycle"
  };
}
function createMetadataDescriptionColumn() {
  return {
    title: "Description",
    field: "entity.metadata.description",
    render: ({entity}) => /* @__PURE__ */ React.createElement(OverflowTooltip, {
      text: entity.metadata.description,
      placement: "bottom-start"
    }),
    width: "auto"
  };
}
function createTagsColumn() {
  return {
    title: "Tags",
    field: "entity.metadata.tags",
    cellStyle: {
      padding: "0px 16px 0px 20px"
    },
    render: ({entity}) => /* @__PURE__ */ React.createElement(React.Fragment, null, entity.metadata.tags && entity.metadata.tags.map((t) => /* @__PURE__ */ React.createElement(Chip, {
      key: t,
      label: t,
      size: "small",
      variant: "outlined",
      style: {marginBottom: "0px"}
    })))
  };
}

var columnFactories = /*#__PURE__*/Object.freeze({
  __proto__: null,
  createNameColumn: createNameColumn,
  createSystemColumn: createSystemColumn,
  createOwnerColumn: createOwnerColumn,
  createSpecTypeColumn: createSpecTypeColumn,
  createSpecLifecycleColumn: createSpecLifecycleColumn,
  createMetadataDescriptionColumn: createMetadataDescriptionColumn,
  createTagsColumn: createTagsColumn
});

const defaultColumns = [
  createNameColumn(),
  createSystemColumn(),
  createOwnerColumn(),
  createSpecTypeColumn(),
  createSpecLifecycleColumn(),
  createMetadataDescriptionColumn(),
  createTagsColumn()
];
const CatalogTable = ({columns}) => {
  var _a, _b;
  const {isStarredEntity, toggleStarredEntity} = useStarredEntities();
  const {loading, error, entities, filters} = useEntityListProvider();
  const showTypeColumn = filters.type === void 0;
  const titlePreamble = capitalize((_b = (_a = filters.user) == null ? void 0 : _a.value) != null ? _b : "all");
  if (error) {
    return /* @__PURE__ */ React.createElement("div", null, /* @__PURE__ */ React.createElement(WarningPanel, {
      severity: "error",
      title: "Could not fetch catalog entities."
    }, /* @__PURE__ */ React.createElement(CodeSnippet, {
      language: "text",
      text: error.toString()
    })));
  }
  const actions = [
    ({entity}) => {
      const url = getEntityMetadataViewUrl(entity);
      return {
        icon: () => /* @__PURE__ */ React.createElement(OpenInNew, {
          fontSize: "small"
        }),
        tooltip: "View",
        disabled: !url,
        onClick: () => {
          if (!url)
            return;
          window.open(url, "_blank");
        }
      };
    },
    ({entity}) => {
      const url = getEntityMetadataEditUrl(entity);
      return {
        icon: () => /* @__PURE__ */ React.createElement(EditIcon, {
          fontSize: "small"
        }),
        tooltip: "Edit",
        disabled: !url,
        onClick: () => {
          if (!url)
            return;
          window.open(url, "_blank");
        }
      };
    },
    ({entity}) => {
      const isStarred = isStarredEntity(entity);
      return {
        cellStyle: {paddingLeft: "1em"},
        icon: () => favouriteEntityIcon(isStarred),
        tooltip: favouriteEntityTooltip(isStarred),
        onClick: () => toggleStarredEntity(entity)
      };
    }
  ];
  const rows = entities.map((entity) => {
    const partOfSystemRelations = getEntityRelations(entity, RELATION_PART_OF, {
      kind: "system"
    });
    const ownedByRelations = getEntityRelations(entity, RELATION_OWNED_BY);
    return {
      entity,
      resolved: {
        name: formatEntityRefTitle(entity, {
          defaultKind: "Component"
        }),
        ownedByRelationsTitle: ownedByRelations.map((r) => formatEntityRefTitle(r, {defaultKind: "group"})).join(", "),
        ownedByRelations,
        partOfSystemRelationTitle: partOfSystemRelations.map((r) => formatEntityRefTitle(r, {
          defaultKind: "system"
        })).join(", "),
        partOfSystemRelations
      }
    };
  });
  const typeColumn = (columns || defaultColumns).find((c) => c.title === "Type");
  if (typeColumn) {
    typeColumn.hidden = !showTypeColumn;
  }
  return /* @__PURE__ */ React.createElement(Table, {
    isLoading: loading,
    columns: columns || defaultColumns,
    options: {
      paging: true,
      pageSize: 20,
      actionsColumnIndex: -1,
      loadingType: "linear",
      showEmptyDataSourceMessage: !loading,
      padding: "dense",
      pageSizeOptions: [20, 50, 100]
    },
    title: `${titlePreamble} (${entities.length})`,
    data: rows,
    actions
  });
};
CatalogTable.columns = columnFactories;

const EntitySwitchCase = (_) => null;
function createSwitchCasesFromChildren(childrenNode) {
  return Children.toArray(childrenNode).flatMap((child) => {
    if (!isValidElement(child)) {
      return [];
    }
    if (child.type === Fragment) {
      return createSwitchCasesFromChildren(child.props.children);
    }
    if (child.type !== EntitySwitchCase) {
      throw new Error(`Child of EntitySwitch is not an EntitySwitch.Case`);
    }
    const {if: condition, children} = child.props;
    return [{if: condition, children}];
  });
}
const EntitySwitch = ({children}) => {
  var _a;
  const {entity} = useEntity();
  const switchCases = useMemo(() => createSwitchCasesFromChildren(children), [
    children
  ]);
  const matchingCase = switchCases.find((switchCase) => switchCase.if ? switchCase.if(entity) : true);
  return (_a = matchingCase == null ? void 0 : matchingCase.children) != null ? _a : null;
};
EntitySwitch.Case = EntitySwitchCase;

function strCmp(a, b) {
  return Boolean(a && (a == null ? void 0 : a.toLowerCase()) === (b == null ? void 0 : b.toLowerCase()));
}
function isKind(kind) {
  return (entity) => strCmp(entity == null ? void 0 : entity.kind, kind);
}
function isComponentType(type) {
  return (entity) => {
    if (!strCmp(entity == null ? void 0 : entity.kind, "component")) {
      return false;
    }
    const componentEntity = entity;
    return strCmp(componentEntity.spec.type, type);
  };
}
function isNamespace(namespace) {
  return (entity) => {
    var _a;
    return strCmp((_a = entity == null ? void 0 : entity.metadata) == null ? void 0 : _a.namespace, namespace);
  };
}

const CatalogLayout = ({children}) => {
  var _a;
  const orgName = (_a = useApi(configApiRef).getOptionalString("organization.name")) != null ? _a : "Backstage";
  return /* @__PURE__ */ React.createElement(Page, {
    themeId: "home"
  }, /* @__PURE__ */ React.createElement(Header, {
    title: `${orgName} Catalog`,
    subtitle: `Catalog of software components at ${orgName}`,
    pageTitleOverride: "Home"
  }), children);
};

const createComponentRouteRef = createExternalRouteRef({
  id: "create-component",
  optional: true
});

const CreateComponentButton = () => {
  const createComponentLink = useRouteRef(createComponentRouteRef);
  if (!createComponentLink)
    return null;
  return /* @__PURE__ */ React.createElement(Button, {
    component: Link$1,
    variant: "contained",
    color: "primary",
    to: createComponentLink()
  }, "Create Component");
};

const useStyles$5 = makeStyles((theme) => ({
  contentWrapper: {
    display: "grid",
    gridTemplateAreas: "'filters' 'table'",
    gridTemplateColumns: "250px 1fr",
    gridColumnGap: theme.spacing(2)
  },
  buttonSpacing: {
    marginLeft: theme.spacing(2)
  }
}));
const CatalogPage = ({
  initiallySelectedFilter = "owned",
  columns
}) => {
  const styles = useStyles$5();
  return /* @__PURE__ */ React.createElement(CatalogLayout, null, /* @__PURE__ */ React.createElement(Content, null, /* @__PURE__ */ React.createElement(ContentHeader, {
    title: "Components"
  }, /* @__PURE__ */ React.createElement(CreateComponentButton, null), /* @__PURE__ */ React.createElement(SupportButton, null, "All your software catalog entities")), /* @__PURE__ */ React.createElement("div", {
    className: styles.contentWrapper
  }, /* @__PURE__ */ React.createElement(EntityListProvider, null, /* @__PURE__ */ React.createElement("div", null, /* @__PURE__ */ React.createElement(EntityKindPicker, {
    initialFilter: "component",
    hidden: true
  }), /* @__PURE__ */ React.createElement(EntityTypePicker, null), /* @__PURE__ */ React.createElement(UserListPicker, {
    initialFilter: initiallySelectedFilter
  }), /* @__PURE__ */ React.createElement(EntityTagPicker, null)), /* @__PURE__ */ React.createElement(CatalogTable, {
    columns
  })))));
};

const EntityLoaderProvider = ({children}) => {
  const {entity, loading, error} = useEntityFromUrl();
  return /* @__PURE__ */ React.createElement(EntityContext.Provider, {
    value: {entity, loading, error}
  }, children);
};

const useStyles$6 = makeStyles((theme) => ({
  illo: {
    maxWidth: "60%",
    top: 100,
    right: 20,
    position: "absolute",
    [theme.breakpoints.down("xs")]: {
      maxWidth: "96%",
      position: "relative",
      top: "unset",
      right: "unset",
      margin: `${theme.spacing(10)}px auto ${theme.spacing(4)}px`
    }
  }
}));
const Illo = () => {
  const classes = useStyles$6();
  return /* @__PURE__ */ React.createElement("img", {
    src: IlloSvgUrl,
    className: classes.illo,
    alt: "Illustration on entity not found page"
  });
};

const useStyles$7 = makeStyles$1((theme) => ({
  container: {
    paddingTop: theme.spacing(24),
    paddingLeft: theme.spacing(8),
    [theme.breakpoints.down("xs")]: {
      padding: theme.spacing(2)
    }
  },
  title: {
    paddingBottom: theme.spacing(2),
    [theme.breakpoints.down("xs")]: {
      fontSize: 32
    }
  },
  body: {
    paddingBottom: theme.spacing(6),
    [theme.breakpoints.down("xs")]: {
      paddingBottom: theme.spacing(5)
    }
  }
}));
const EntityNotFound = () => {
  const classes = useStyles$7();
  return /* @__PURE__ */ React.createElement(Grid, {
    container: true,
    spacing: 0,
    className: classes.container
  }, /* @__PURE__ */ React.createElement(Illo, null), /* @__PURE__ */ React.createElement(Grid, {
    item: true,
    xs: 12,
    sm: 6
  }, /* @__PURE__ */ React.createElement(Typography, {
    variant: "h2",
    className: classes.title
  }, "Entity was not found"), /* @__PURE__ */ React.createElement(Typography, {
    variant: "body1",
    className: classes.body
  }, "Want to help us build this? Check out our Getting Started documentation."), /* @__PURE__ */ React.createElement(Button, {
    variant: "contained",
    color: "primary",
    href: "https://backstage.io/docs"
  }, "DOCS")));
};

const DefaultEntityPage = () => /* @__PURE__ */ React.createElement(EntityPageLayout, null, /* @__PURE__ */ React.createElement(EntityPageLayout.Content, {
  path: "/",
  title: "Overview",
  element: /* @__PURE__ */ React.createElement(Content, null, /* @__PURE__ */ React.createElement(Typography, {
    variant: "h2"
  }, "This is the default entity page."), /* @__PURE__ */ React.createElement(Typography, {
    variant: "body1"
  }, "To override this component with your custom implementation, read docs on", " ", /* @__PURE__ */ React.createElement(Link$2, {
    target: "_blank",
    href: "https://backstage.io/docs"
  }, "backstage.io/docs")))
}));
const EntityPageSwitch = ({EntityPage}) => {
  const {entity, loading, error} = useEntity();
  if (loading)
    return /* @__PURE__ */ React.createElement(EntityPageLayout, null);
  if (error || !entity)
    return /* @__PURE__ */ React.createElement(EntityNotFound, null);
  return /* @__PURE__ */ React.createElement(EntityPage, null);
};
const OldEntityRouteRedirect = () => {
  var _a;
  const {optionalNamespaceAndName, "*": rest} = useParams();
  const [name, namespace] = optionalNamespaceAndName.split(":").reverse();
  const namespaceLower = (_a = namespace == null ? void 0 : namespace.toLocaleLowerCase("en-US")) != null ? _a : ENTITY_DEFAULT_NAMESPACE;
  const restWithSlash = rest ? `/${rest}` : "";
  return /* @__PURE__ */ React.createElement(Navigate, {
    to: `../../${namespaceLower}/component/${name}${restWithSlash}`
  });
};
const Router = ({
  EntityPage = DefaultEntityPage
}) => /* @__PURE__ */ React.createElement(Routes, null, /* @__PURE__ */ React.createElement(Route$1, {
  path: `${rootRoute.path}`,
  element: /* @__PURE__ */ React.createElement(CatalogPage, null)
}), /* @__PURE__ */ React.createElement(Route$1, {
  path: `${entityRoute.path}`,
  element: /* @__PURE__ */ React.createElement(EntityLoaderProvider, null, /* @__PURE__ */ React.createElement(EntityPageSwitch, {
    EntityPage
  }))
}), /* @__PURE__ */ React.createElement(Route$1, {
  path: "Component/:optionalNamespaceAndName/*",
  element: /* @__PURE__ */ React.createElement(OldEntityRouteRedirect, null)
}));

class CatalogClientWrapper {
  constructor(options) {
    this.client = options.client;
    this.identityApi = options.identityApi;
  }
  async getLocationById(id, options) {
    var _a;
    return await this.client.getLocationById(id, {
      token: (_a = options == null ? void 0 : options.token) != null ? _a : await this.identityApi.getIdToken()
    });
  }
  async getEntities(request, options) {
    var _a;
    return await this.client.getEntities(request, {
      token: (_a = options == null ? void 0 : options.token) != null ? _a : await this.identityApi.getIdToken()
    });
  }
  async getEntityByName(compoundName, options) {
    var _a;
    return await this.client.getEntityByName(compoundName, {
      token: (_a = options == null ? void 0 : options.token) != null ? _a : await this.identityApi.getIdToken()
    });
  }
  async addLocation(request, options) {
    var _a;
    return await this.client.addLocation(request, {
      token: (_a = options == null ? void 0 : options.token) != null ? _a : await this.identityApi.getIdToken()
    });
  }
  async getOriginLocationByEntity(entity, options) {
    var _a;
    return await this.client.getOriginLocationByEntity(entity, {
      token: (_a = options == null ? void 0 : options.token) != null ? _a : await this.identityApi.getIdToken()
    });
  }
  async getLocationByEntity(entity, options) {
    var _a;
    return await this.client.getLocationByEntity(entity, {
      token: (_a = options == null ? void 0 : options.token) != null ? _a : await this.identityApi.getIdToken()
    });
  }
  async removeLocationById(id, options) {
    var _a;
    return await this.client.removeLocationById(id, {
      token: (_a = options == null ? void 0 : options.token) != null ? _a : await this.identityApi.getIdToken()
    });
  }
  async removeEntityByUid(uid, options) {
    var _a;
    return await this.client.removeEntityByUid(uid, {
      token: (_a = options == null ? void 0 : options.token) != null ? _a : await this.identityApi.getIdToken()
    });
  }
}

const catalogPlugin = createPlugin({
  id: "catalog",
  apis: [
    createApiFactory({
      api: catalogApiRef,
      deps: {discoveryApi: discoveryApiRef, identityApi: identityApiRef},
      factory: ({discoveryApi, identityApi}) => new CatalogClientWrapper({
        client: new CatalogClient({discoveryApi}),
        identityApi
      })
    })
  ],
  routes: {
    catalogIndex: catalogRouteRef,
    catalogEntity: entityRouteRef
  },
  externalRoutes: {
    createComponent: createComponentRouteRef
  }
});
const CatalogIndexPage = catalogPlugin.provide(createRoutableExtension({
  component: () => import('./index-6656a371.esm.js').then((m) => m.CatalogPage),
  mountPoint: catalogRouteRef
}));
const CatalogEntityPage = catalogPlugin.provide(createRoutableExtension({
  component: () => import('./index-2362b904.esm.js').then((m) => m.CatalogEntityPage),
  mountPoint: entityRouteRef
}));
const EntityAboutCard = catalogPlugin.provide(createComponentExtension({
  component: {
    lazy: () => import('./index-6f202e28.esm.js').then((m) => m.AboutCard)
  }
}));
const EntityLinksCard = catalogPlugin.provide(createComponentExtension({
  component: {
    lazy: () => import('./index-e628ccf2.esm.js').then((m) => m.EntityLinksCard)
  }
}));
const EntityHasSystemsCard = catalogPlugin.provide(createComponentExtension({
  component: {
    lazy: () => import('./index-4ef77e53.esm.js').then((m) => m.HasSystemsCard)
  }
}));
const EntityHasComponentsCard = catalogPlugin.provide(createComponentExtension({
  component: {
    lazy: () => import('./index-b3e21eee.esm.js').then((m) => m.HasComponentsCard)
  }
}));
const EntityHasSubcomponentsCard = catalogPlugin.provide(createComponentExtension({
  component: {
    lazy: () => import('./index-3ace567b.esm.js').then((m) => m.HasSubcomponentsCard)
  }
}));
const EntityHasResourcesCard = catalogPlugin.provide(createComponentExtension({
  component: {
    lazy: () => import('./index-593342f8.esm.js').then((m) => m.HasResourcesCard)
  }
}));
const EntityDependsOnComponentsCard = catalogPlugin.provide(createComponentExtension({
  component: {
    lazy: () => import('./index-36b66f8e.esm.js').then((m) => m.DependsOnComponentsCard)
  }
}));
const EntityDependsOnResourcesCard = catalogPlugin.provide(createComponentExtension({
  component: {
    lazy: () => import('./index-d231361f.esm.js').then((m) => m.DependsOnResourcesCard)
  }
}));
const EntitySystemDiagramCard = catalogPlugin.provide(createComponentExtension({
  component: {
    lazy: () => import('./index-d159831b.esm.js').then((m) => m.SystemDiagramCard)
  }
}));

const DeleteEntityDialog = ({
  open,
  onClose,
  onConfirm,
  entity
}) => {
  const [busy, setBusy] = useState(false);
  const catalogApi = useApi(catalogApiRef);
  const alertApi = useApi(alertApiRef);
  const onDelete = async () => {
    setBusy(true);
    try {
      const uid = entity.metadata.uid;
      await catalogApi.removeEntityByUid(uid);
      onConfirm();
    } catch (err) {
      alertApi.post({message: err.message});
    } finally {
      setBusy(false);
    }
  };
  return /* @__PURE__ */ React.createElement(Dialog, {
    open,
    onClose
  }, /* @__PURE__ */ React.createElement(DialogTitle, {
    id: "responsive-dialog-title"
  }, "Are you sure you want to delete this entity?"), /* @__PURE__ */ React.createElement(DialogActions, null, /* @__PURE__ */ React.createElement(Button, {
    variant: "contained",
    color: "secondary",
    disabled: busy,
    onClick: onDelete
  }, "Delete"), /* @__PURE__ */ React.createElement(Button, {
    onClick: onClose,
    color: "primary"
  }, "Cancel")));
};

const isOrphan = (entity) => {
  var _a, _b;
  return ((_b = (_a = entity == null ? void 0 : entity.metadata) == null ? void 0 : _a.annotations) == null ? void 0 : _b["backstage.io/orphan"]) === "true";
};
const EntityOrphanWarning = () => {
  const navigate = useNavigate();
  const catalogLink = useRouteRef(catalogRouteRef);
  const [confirmationDialogOpen, setConfirmationDialogOpen] = useState(false);
  const {entity} = useEntity();
  const cleanUpAfterRemoval = async () => {
    setConfirmationDialogOpen(false);
    navigate(catalogLink());
  };
  return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(Alert$1, {
    severity: "warning",
    onClick: () => setConfirmationDialogOpen(true)
  }, "This entity is not referenced by any location and is therefore not receiving updates. Click here to delete."), /* @__PURE__ */ React.createElement(DeleteEntityDialog, {
    open: confirmationDialogOpen,
    entity,
    onConfirm: cleanUpAfterRemoval,
    onClose: () => setConfirmationDialogOpen(false)
  }));
};

export { AboutCard as A, CatalogPage as C, EntityLoaderProvider as E, Router as R, EntityLayout as a, EntityPageLayout as b, CatalogTable as c, CatalogEntityPage as d, CatalogIndexPage as e, catalogPlugin as f, EntityAboutCard as g, EntityDependsOnComponentsCard as h, EntityDependsOnResourcesCard as i, EntityHasComponentsCard as j, EntityHasResourcesCard as k, EntityHasSubcomponentsCard as l, EntityHasSystemsCard as m, EntityLinksCard as n, EntitySystemDiagramCard as o, EntitySwitch as p, isKind as q, isNamespace as r, isComponentType as s, EntityOrphanWarning as t, isOrphan as u };
